cpp/include/rapids_triton/model/shared_state.hpp

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+#include <cuda_runtime_api.h>
+#include <algorithm>
+#include <cstddef>
+#include <optional>
+#include <sstream>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include <triton/backend/backend_common.h>
+#include <rapids_triton/batch/batch.hpp>
+#include <rapids_triton/tensor/tensor.hpp>
+#include <rapids_triton/triton/config.hpp>
+#include <rapids_triton/triton/deployment.hpp>
+#include <rapids_triton/utils/narrow.hpp>
+
+namespace triton {
+namespace backend {
+namespace rapids {
+/**
+ * @brief Stores shared state for multiple instances of the same model
+ */
+struct SharedModelState {
+  virtual void load() {}
+  virtual void unload() {}
+
+  explicit SharedModelState(std::unique_ptr<common::TritonJson::Value>&& config)
+    : config_{std::move(config)},
```

Comment:
Is the `std::move` here unnecessary? Since it's not a deduced template param, I would imagine the constructor would only take r-values of `config` instead of u-values, but I could be wrong

Reply:
It is because `config` itself is an lvalue (it has a name and duration) even though an rvalue reference was passed into it. Here's a little snippet to prove it to yourself:
```cpp
#include <iostream>

struct A {
  A() { std::cout << "DEFAULT!\n"; }
  A(A const& in) { std::cout << "COPY!\n"; }
  A(A&& in) { std::cout << "MOVE!\n"; }
};

struct B {
  B(A&& in) : member{std::move(in)} {}

  A member;
};

int main() {
  auto b = B(A());
  return 0;
}
```

If you compile that (without optimizations) and run it, you'll get `MOVE!` in the output, but if you remove the `std::move`, you'll get `COPY!`.